app/Http/Controllers/Api/DeviceController.php

```diff
@@ -318,5 +322,69 @@ public function anyList(BcsRequest $request)
 
         return response()->json($ret_val);
     }
-
+    
+    /**
+     * @param \BComeSafe\Http\Requests\SaveDeviceRequest $request
+     *
+     * @return array
+     */
+    public function anyUpdateDevice(SaveDeviceRequest $request)
+    {
+        $update = array();
+        $params = $request->all();
+        //
+        if (isset($params['user_phone'])) {
+          if (!empty($request->get('user_phone'))) {
+            $update['user_phone'] = $request->get('user_phone');
+            $update['need_phone'] = 0;
+            $update['user_phone_confirm'] = 0;
+            //Generate token
+            $update['user_phone_token'] = Device::generateToken();
+            //Send token via SMS
+            $defaults = SchoolDefault::getDefaults();
+            if (!empty($defaults->sms_provider)) {
+              $integration = \Component::get('Sms')->getIntegration($defaults->sms_provider);
+              $integration->sendMessage($update['user_phone'], $update['user_phone_token']);
+            }
+          }
+          else {
+            $update['user_phone'] = '';
+            $update['need_phone'] = 1;
+            $update['user_phone_confirm'] = 0;
+            $update['user_phone_token'] = '';
+          }
+          
+        }
+        //
+        if (!empty($request->get('skip_phone'))) {
+          $update['need_phone'] = 0;
+        }
+        //
+        if (!empty($request->get('user_phone_token'))) {
+          $device_data = Device::getByDeviceId($request->get('device_id'));
+          if (!empty($device_data['user_phone_token']) && $device_data['user_phone_token'] == $request->get('user_phone_token')) {
+            $update['user_phone_confirm'] = 1;
+          }
+        }
+        //
+        if (isset($params['accepted_tac'])) {
+          if (!empty($request->get('accepted_tac'))) {
```

Comment:
Could probably also be replaced by `if ($request->filled ...`

Reply:
Todo.